### PR TITLE
convert project code to typescript.

### PR DESCRIPTION
**Pull request by UrielCh**

### CODE REVIEW

File: demo/shared/views/create-group-boxes.js [hunk 1]
```diff
@@ -73,7 +73,7 @@ module.exports = function ({mean, covariance, color, parent, className, tag = 'd
 		color,
 	});
 	const arrowRotation = (-1 * Math.atan(mean[4][0] / mean[5][0]) * 180 / Math.PI) - 45;
-	const arrowScale = Math.sqrt((mean[4][0] ** 2) + (mean[5][0] ** 2));
+	const arrowScale = Math.hypot((mean[4][0]), (mean[5][0]));
```
Comment (UrielCh) [inline demo/shared/views/create-group-boxes.js:76]: extract form hypot doc:
```js
 /**
     * Returns the square root of the sum of squares of its arguments.
     * @param values Values to compute the square root for.
     *     If no arguments are passed, the result is +0.
     *     If there is only one argument, the result is the absolute value.
     *     If any argument is +Infinity or -Infinity, the result is +Infinity.
     *     If any argument is NaN, the result is NaN.
     *     If all arguments are either +0 or −0, the result is +0.
     */
```

File: index.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import * as modelCollection from './lib/model-collection';
+import * as defaultDynamicModels from './lib/dynamic';
+import * as defaultObservationModels from './lib/observation';
+
+function camelToDash(str: string) {
```
Comment (UrielCh) [inline index.ts:5]: Original export are named using Dash syntax, but this syntax cause export issues, so intread of exporting `my-dynamic-model` I export `myDynamicModel` the convert it back to the Dash syntax to keep the existing syntax.

File: lib/core-kalman-filter.ts [hunk 3]
```diff
@@ -80,19 +49,19 @@ class CoreKalmanFilter {
...
 	*/
 
-	getPredictedCovariance(options = {}) {
+	getPredictedCovariance(options: {previousCorrected?: State, index?: number} = {}) {
 		let {previousCorrected, index} = options;
-		previousCorrected = previousCorrected || this.getInitState();
+		previousCorrected ||= this.getInitState();
 
-		const getValueOptions = Object.assign({}, {previousCorrected, index}, options);
+		const getValueOptions = {previousCorrected, index, ...options};
```
Comment (UrielCh) [inline lib/core-kalman-filter.ts:56]: I replace all Object.assign by spread opperator, spread opperator maintain types, Object.assign break them.

spread opperator can be transpile as Object.assign if needed.

File: demo/bike/main.js [hunk 4]
```diff
@@ -1,13 +1,17 @@
-const {KalmanFilter} = kalmanFilter;// eslint-disable-line no-undef
-const createElement = require('../shared/views/create-element');
+/**
+ * @typedef {typeof import('../../index.d.ts')} kalmanFilter
+ */
+const { KalmanFilter } = /** @type {kalmanFilter} */ (kalmanFilter);// eslint-disable-line no-undef
```
Comment (UrielCh) [inline demo/bike/main.js:4]: now `kalmanFilter` is properly typed with all the lib typing withing this JS file.

File: demo/bike/main.js [hunk 5]
```diff
@@ -1,13 +1,17 @@
...
+const createElement = require('../shared/views/create-elem§ent');
 const createGroupBoxes = require('../shared/views/create-group-boxes');
 const noisyObservations = require('./observations.json').observations;
 const kfOptions = require('./kf-options');
-
 const kf = new KalmanFilter(kfOptions);
```
Comment (UrielCh) [inline demo/bike/main.js:9]: here `kf` is properly typed as a KalmanFilter instance.

File: lib/dynamic/index.ts [hunk 6]
```diff
@@ -0,0 +1,19 @@
+export {default as constantPosition} from './constant-position';
+export {default as constantSpeed} from './constant-speed';
+export {default as constantAcceleration} from './constant-acceleration';
+export {default as composition} from './composition';
+export {default as constantPositionWithNull} from './constant-position-with-null';
+export {default as constantSpeedDynamic} from './constant-speed-dynamic';
+export {default as shorttermConstantSpeed} from './shortterm-constant-speed';
+
+// module.exports = {
```
Comment (UrielCh) [inline lib/dynamic/index.ts:9]: old Dashed syntax replaced by camelCase

File: lib/core-kalman-filter.ts (hunk 3 above)
Comment (piercus) [reply lib/core-kalman-filter.ts:56]: spread operator is ok

File: demo/bike/main.js [hunk 7]
```diff
@@ -1,13 +1,17 @@
-const {KalmanFilter} = kalmanFilter;// eslint-disable-line no-undef
-const createElement = require('../shared/views/create-element');
+/**
+ * @typedef {typeof import('../../index.d.ts')} kalmanFilter
```
[the remark this answers is not in the document]
Comment (piercus) [reply demo/bike/main.js:2]: ok

File: demo/bike/main.js (hunk 4 above)
Comment (piercus) [reply demo/bike/main.js:4]: ok

File: demo/bike/main.js (hunk 5 above)
Comment (piercus) [reply demo/bike/main.js:9]: ok

File: pnpm-lock.yaml
Comment (piercus) [inline pnpm-lock.yaml:1]: please remove this file from git

**Review by piercus — CHANGES_REQUESTED** [review]
Thanks for this work, very useful

Can you please also move the PR to piercus:develop branch ?

Thanks

**Comment by UrielCh** [comment]
I can only merge it to an existing branch.
and I do not see any existing `develop` branch

**Comment by piercus** [comment]
@UrielCh thanks for your work

Can you please squash all your commits into one "BREAKING CHANGE" commit (so it will trigger a new Major release) see https://github.com/semantic-release/semantic-release

**Comment by piercus** [comment]
Please trigger a major release for this, if there is no breaking change this is a good point

I feel it's quite possible that some CI tool, static checker ... in consuming libraries might be impacted by this ts release.

Let's be safe

**Comment by UrielCh** [comment]
Ok

**Comment by UrielCh** [comment]
Check PR #75

**Pull request closed without merging**